File: prince-of-versions/src/main/java/co/infinum/princeofversions/UpdaterCall.java
package co.infinum.princeofversions;

import java.io.IOException;
import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicBoolean;

import javax.annotation.Nullable;

/**
 * An invocation of a {@link PrinceOfVersions} update check method.
 * The instance can be used only once, if call is already executed or enqueued new instnace should be used instead.
 *
 * <p>Calls may be executed synchronously with {@link #execute}, or asynchronously with {@link #enqueue}.
 * In either case the call can be canceled at any time with {@link #cancel}.
 */
public final class UpdaterCall implements PrinceOfVersionsCall {

    private final PrinceOfVersions core;
    private final Loader loader;

    private final AtomicBoolean executed = new AtomicBoolean(false);
    private final AtomicBoolean canceled = new AtomicBoolean(false);

    @Nullable
    private PrinceOfVersionsCancelable cancelable;

    UpdaterCall(final PrinceOfVersions core, final Loader loader) {
        this.core = core;
        this.loader = loader;
    }

    @Override
    public UpdateResult execute() throws Throwable {
        if (executed.getAndSet(true)) {
            throw new IllegalStateException("Already executed!");
        }
        if (canceled.get()) {
            throw new IOException("Canceled!");
        }
        return core.checkForUpdates(loader);
    }

    @Override
    public void enqueue(final UpdaterCallback callback) {
        if (executed.getAndSet(true)) {
            throw new IllegalStateException("Already executed!");
        }
        if (canceled.get()) {
            callback.onError(new IOException("Canceled"));
        }
        cancelable = core.checkForUpdates(loader, callback);
    }

    @Override
    public void enqueue(final Executor executor, final UpdaterCallback callback) {
        if (executed.getAndSet(true)) {
            throw new IllegalStateException("Already executed!");
        }
        if (canceled.get()) {
            callback.onError(new IOException("Canceled"));
        }
        cancelable = core.checkForUpdates(executor, loader, callback);
    }

    @Override
    public void cancel() {
        canceled.set(true);

        if (cancelable != null) {
            cancelable.cancel();
        }
    }

    @Override
    public boolean isCanceled() {
        return canceled.get();
    }
}

File: prince-of-versions/src/main/java/co/infinum/princeofversions/UpdaterCall.kt
package co.infinum.princeofversions

import java.io.IOException
import java.util.concurrent.Executor
import java.util.concurrent.atomic.AtomicBoolean

/**
 * An invocation of a [PrinceOfVersions] update check method.
 * The instance can be used only once; if a call is already executed or enqueued, a new instance should be used instead.
 *
 * Calls may be executed synchronously with [execute] or asynchronously with [enqueue].
 * In either case, the call can be canceled at any time with [cancel].
 */
internal class UpdaterCall(
    private val core: PrinceOfVersions,
    private val loader: Loader
) : PrinceOfVersionsCall {

    companion object {
        private const val CANCELED_MESSAGE = "Canceled!"
        private const val ALREADY_EXECUTED_MESSAGE = "Already executed!"
    }

    private val executed = AtomicBoolean(false)
    private val canceled = AtomicBoolean(false)

    private var cancelable: PrinceOfVersionsCancelable? = null

    @Throws(Throwable::class)
    override fun execute(): UpdateResult {
        check(!(executed.getAndSet(true))) { ALREADY_EXECUTED_MESSAGE }
        if (isCanceled) {
            throw IOException(CANCELED_MESSAGE)
        }
        return core.checkForUpdates(loader)
    }

    override fun enqueue(callback: UpdaterCallback) {
        check(!(executed.getAndSet(true))) { ALREADY_EXECUTED_MESSAGE }
        if (isCanceled) {
            callback.onError(IOException(CANCELED_MESSAGE))
            return
        }
        cancelable = core.checkForUpdates(loader, callback)
    }

    override fun enqueue(executor: Executor, callback: UpdaterCallback) {
        check(!(executed.getAndSet(true))) { ALREADY_EXECUTED_MESSAGE }
        if (isCanceled) {
            callback.onError(IOException(CANCELED_MESSAGE))
            return
        }
        cancelable = core.checkForUpdates(executor, loader, callback)
    }

    override fun cancel() {
        canceled.set(true)
        cancelable?.cancel()
    }

    override val isCanceled: Boolean
        get() = canceled.get()
}